Userland/Libraries/LibAudio/PlaybackStreamOboe.cpp

```diff
@@ -0,0 +1,158 @@
+/*
+ * Copyright (c) 2024, Olekoop <mlglol360xd@gmail.com>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#define AK_DONT_REPLACE_STD
+#include <AK/Atomic.h>
+#include <AK/SourceLocation.h>
+#include <LibAudio/PlaybackStreamOboe.h>
+#include <LibCore/SharedCircularQueue.h>
+#include <LibCore/ThreadedPromise.h>
+#include <memory>
+
+#include <oboe/Oboe.h>
+
+namespace Audio {
+
+class OboeCallback : public oboe::AudioStreamDataCallback {
+public:
+    virtual oboe::DataCallbackResult onAudioReady(oboe::AudioStream* oboeStream, void* audioData, int32_t numFrames)
+    {
+        Bytes output_buffer {
+            reinterpret_cast<u8*>(audioData),
+            static_cast<size_t>(numFrames * oboeStream->getChannelCount() * sizeof(float))
+        };
+        auto written_bytes = m_data_request_callback(output_buffer, PcmSampleFormat::Float32, numFrames);
+        if (written_bytes.is_empty())
+            return oboe::DataCallbackResult::Stop;
+
+        auto timestamp = oboeStream->getTimestamp(CLOCK_MONOTONIC);
+        if (timestamp == oboe::Result::OK) {
+            m_number_of_samples_enqueued = timestamp.value().position;
+        } else {
+            // Fallback for OpenSLES
+            m_number_of_samples_enqueued += numFrames;
+        }
+        auto last_sample_time = static_cast<i64>(m_number_of_samples_enqueued / oboeStream->getSampleRate());
+        m_last_sample_time.store(last_sample_time);
```

Comment:
This doesn't seem right to me, wouldn't the value returned by `getTimestamp()` be in some time units rather than number of samples? I can't find in the docs if they mention the units, but it would be nice to confirm that this works correctly.

I'm also curious about the "Fallback for OpenSLES" comment, why is this necessary? Does this comment mean that one of these two code paths is untested?

Reply:
Well, getTimestamp() returns a struct with both a number of samples and a time in nanoseconds, however the time seems to be counted from Unix epoch, not from the start of playing, so I used the number of samples.

The getTimestamp() is not implemented in Oboe when using OpenSLES (this happens when AAudio is not available), therefore I implemented this fallback. I tested both of these and they work. The fallback will not probably be ever used, since AAudio is available since Android 8 and the Ladybird requires at least Android 11.

Reply:
Ah, right, I missed that you were using the frame position here, it should be fine in that regard then.

One related concern I have is that the docs mention that the function shouldn't be called from the data callback, so perhaps if the fallback isn't ever required, we can make things a bit simpler and call getTimestamp() directly in total_time_played().

Reply:
I guess it appears that that warning about using it in the callback is related to the OpenSLES fallback you have implemented, personally this assertion manifests as an error in Oboe versions where's it was fixed? So then the question is just whether we have any need to support that case if we're not targeting Android 10?

Reply:
I think that it should be left as is. If Ladybird would like to have older Android support, it would make the job easier.

Reply:
I'll leave that up to @ADKaster, I think he would know more about what we should do here. It's worth noting that getTimestamp() is likely to change values more often than our data callback is called, especially if it interpolates the actual time from the output device, whereas the data callback's timestamps would only increment by a specific interval determined by the frequency that it is called in.

Reply:
Big Meh on that. Going below API level 30 will require extra legwork in LibCore to backfill missing Bionic APIs. But if it's not hurting anything it seems fine to go with it for this first implementation.